DOCS-554-601: revise frame system page (#1007)
viamrobotics/docs: docs/services/frame-system/_index.md

---
title: "The Robot Frame System"
linkTitle: "Frame System"
description: "The Frame System holds reference frame information for the relative position of components in space."
type: docs
weight: 45
no_list: true
tags: ["frame system", "services"]
# SMEs: Peter L, Gautham, Bijan
---

> Any robot configured in Viam comes with a service we call the Frame System: an internally managed and mostly static system for storing the "reference frame" of each component of a robot within a coordinate system configured by the user.

Author: Ah yes, done!

> The Frame System is the basis for many of Viam's other services, like [motion](/services/motion) and [vision](/services/vision).

Reviewer: ```suggestion
The Frame System is the basis for many of Viam's other services, like [Motion](/services/motion) and [Vision](/services/vision).
```

Author: Done!

It stores the required contextual information to use the position and orientation readings returned by some components.

## Configuration

To adjust the Frame System from its default configuration for a particular [component](/components), add the following to its configuration:

{{< tabs name="Frame Configuration Instructions" >}}
{{% tab name="Config Builder" %}}

> Navigate to the **CONFIG** tab on your robot's page in [the Viam app](https://app.viam.com), select the **Builder** mode, scroll to a component's card, and click **Add Frame**:

Author: @npentrel Tried to address on line 28 below, LMK what you think

        "orientation": {
          "type": <"ov_degrees">,
          "value": {
            "x": 0,
            "y": 0,
            "z": 1,
            "th": 0
          }
        }
      }
    }
  ]
}
```

{{% /tab %}}
{{< /tabs >}}

Configure the reference frame as follows:

| Parameter | Inclusion | Required |
| --------- | ----------- | ----- |
| `Parent`  | **Required** | Default: `world`. The name of the reference frame you want to act as the parent of this frame. |
| `Translation` | **Required** | Default: `(0, 0, 0)`. The coordinates that the origin of this component's reference frame has within its parent reference frame. |
| `Orientation`  | **Required** | Default: `(0, 0, 1), 0`. The [orientation vector](/internals/orientation-vector/) that yields the axes of the component's reference frame when applied as a rotation to the axes of the parent reference frame. <br> Types: `Orientation Vector Degrees`, `Orientation Vector Radians`, and `Quaternion`. |
| `Geometry`  | **Required** | Default: `none`. Collision geometries for defining bounds in the environment of the robot. <br> Types: `Sphere`, `Box`, and `Capsule`. |

{{% alert title="Caution" color="caution" %}}

> `Types` are offered in `Orientation` for ease of configuration, but the [orientation vector](/internals/orientation-vector/) type that is always stored and returned by the Frame System is `"Orientation Vector Radians"`.

Reviewer 2: can we put this caution alert in the table?

`"Degrees"` and `"Quaternion"` will be converted to `"Radians"`.

{{% /alert %}}

{{% alert title="Tip" color="tip" %}}

> Viam's coordinate system considers `+X` to be forward, `+Y` to be left, and `+Z` to be up.

Reviewer: [suggest] Make mention of the "right hand rule"? https://en.wikipedia.org/wiki/Right-hand_rule

Author: done, LMK if you have any suggestions

{{% /alert %}}

For more information, see these two examples:

- [A component attached to a static surface](/component-on-static)
- [A component attached to another, dynamic, component](/component-on-dynamic)

## Building the Frame System

Once you configure your robot and run `viam-server`, your robot builds a tree of reference frames with the `world` as the root node.

> A [topologically-sorted list](https://en.wikipedia.org/wiki/Topological_sorting) of the generated reference frames is printed by the server and can be seen in the logs at `--debug` level:

Reviewer 2: remove passive voice

Author: done!

![an example of a logged frame system](img/frame_sys_log_example.png)

`viam-server` regenerates this tree in the process of [reconfiguration](/manage/fleet-management/#configurationlogging).

Let's consider the example of a [component attached to a dynamic component](/component-attached-to-a-dynamic-component): a robotic arm, `A`, attached to a gantry, `G`, which in turn is fixed in place at a point on the `World` of a table.

The resulting tree of reference frames looks like:

![reference frame tree](img/frame_tree.png)

`viam-server` builds the connections in this tree by looking at the `"frame"` portion of each component in the robot's configuration and defining *two* reference frames for each component:

1. One with the name of the component, representing the actuator or final link in the component's kinematic chain: like `"A"` as the end of an arm.
2. Another representing the origin of the component, defined with the component's name and the suffix *"_origin"*.

## Accessing the Frame System

The [Robot API](https://github.com/viamrobotics/api/blob/main/proto/viam/robot/v1/robot.proto) supplies two methods to interact with the Frame System through gRPC calls:

> 1. `TransformPose`: Transforms a pose measured in one reference frame to the same pose as it would have been measured in another.

Reviewer 2: We should probably have the api methods here like in components

Author: https://viam.atlassian.net/browse/DOCS-674?atlOrigin=eyJpIjoiOTUzNmU1ZmZlYjA0NDAwYWEzZmZiODBmMDk5OTg2YzgiLCJwIjoiaiJ9

2. `FrameSystemConfig`: Returns a topologically sorted list of all the reference frames monitored by the frame system.
Any [supplemental transforms](#handling-motion-with-supplemental-transforms) are also merged into the tree, topologically sorted, and returned.

> ## Supplemental Transforms

Reviewer: I need to read over this section more thoroughly, there is a decent amount of information here that feels outdated and I want to double-check that.

Reviewer: @biotinker and I are going to talk about this together during our morning 1:1, so we should have some updates for you shortly.

In our [example of dynamic attachment](/component-attached-to-a-dynamic-component), the arm can be managed by the Frame System without supplemental transforms because the base of the arm is fixed with respect to the gantry's platform, and the gantry's origin is fixed with respect to the `world` reference frame.

However, a dynamic arm attached to a [dynamic rover](/components/base/wheeled) could not be configured into the Frame System if the rover is unaware of its own position, because the rover can move freely with respect to the `world` reference frame.

To solve this problem:

- You can introduce a [movement sensor](/components/movement-sensor) or a [camera](/components/camera), in combination with our [Vision Service](/services/vision/), as a third component.
- This component would be fixed in respect to the `world` reference frame, and could supply the location and orientation of the rover in its own reference frame.

To add this component's reference frame into your robot's Frame System build, you can pass the `name` of this component's reference frame to various APIs.

> These *supplemental transforms* would be the missing link to be able to transform a [pose](/internals/orientation-vector) in that dynamic arm's reference frame to the `world` reference frame.

Reviewer 2: ```suggestion
These *supplemental transforms* then supply the missing link to transform a [pose](/internals/orientation-vector) in that dynamic arm's reference frame to the `world` reference frame.
```

Author: done!

- For example, you can pass this component's reference frame information to the `supplemental_transforms` parameter in your calls to Viam's motion service [`GetPose`](/services/motion/#getpose) method.
- Functions of some services and components also take in a `WorldState` parameter, which includes a `transforms` parameter.
- Both [`TransformPose`](#accessing-the-frame-system) and [`FrameSystemConfig`](#accessing-the-frame-system) have the option to take in these supplemental transforms.

{{% alert title="Tip" color="tip" %}}

> A knowledgeable user could code a [mobile base](/components/base/wheeled) that is able to report its own position without the need for supplemental transforms with an organic [SLAM](/services/slam) system.

Reviewer 2: Avoid conditional
```suggestion
Experiences users can code a [mobile base](/components/base/wheeled) that is able to report its own position without the need for supplemental transforms with an organic [SLAM](/services/slam) system.
```

| `Geometry`  | Optional | Default: `none`. Collision geometries for defining bounds in the environment of the robot. <br> Types: `Sphere`, `Box`, and `Capsule`. |

{{% alert title="Caution" color="caution" %}}

The `Orientation` parameter offers `Types` for ease of configuration, but the Frame System always stores and returns [orientation vectors](/internals/orientation-vector/) in `"Orientation Vector Radians"`.
`"Degrees"` and `"Quaternion"` will be converted to `"Radians"`.

{{% /alert %}}

{{% alert title="Tip" color="tip" %}}

Viam's coordinate system considers `+X` to be forward, `+Y` to be left, and `+Z` to be up.
You can use [the right-hand rule](https://en.wikipedia.org/wiki/Right-hand_rule) to determine the orientation of these axes.

{{% /alert %}}

For more information about determining the appropriate values for these parameters, see these two examples:

- [A component attached to a static surface](/services/frame-system/component-on-static)
- [A component attached to another, dynamic, component](/services/frame-system/component-on-dynamic)

## Building the Frame System

> Access a [topologically-sorted list](https://en.wikipedia.org/wiki/Topological_sorting) of the generated reference frames in the robot's logs at `--debug` level:

Author: done!

    If a [camera](/components/camera) is looking for an apple or an orange, the arm could be commanded to move to the detected fruit's location by providing a supplemental transform that contains the detected pose with respect to the camera that performed the detection.

    The detecting component (camera) would be fixed with respect to the `world` reference frame, and would supply the position and orientation of the detected object.

    With this information, the Frame System could perform the right calculations to express the pose of the object in the `world` reference frame.

Usage:

- You can pass a detected object's frame information to the `supplemental_transforms` parameter in your calls to Viam's Motion Service's [`GetPose`](/services/motion/#getpose) method.
- Functions of some services and components also take in a `WorldState` parameter, which includes a `transforms` property.
- Both [`TransformPose`](#accessing-the-frame-system) and [`FrameSystemConfig`](#accessing-the-frame-system) have the option to take in these supplemental transforms.

{{% alert title="TBD" color="note" %}}

Additional uses for supplemental transforms are being explored by the Viam team.

> {{% /alert %}}

Author: removed!
